orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,51 +714,95 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
 		if actionEffect != certprotos.Effect_UNKNOWN &&
 			resourceEffect != certprotos.Effect_UNKNOWN &&
 			actionEffect == resourceEffect {
 			effect = actionEffect
 		}
+
+		if actionEffect == certprotos.Effect_DENY &&
+			resourceEffect == certprotos.Effect_DENY {
+			return certprotos.Effect_DENY, nil
+		}
 	}
 
 	return effect, nil
 }
 
-// checkResource checks if the requested resource is authorized by the policy
-func checkResource(reqResource, policyResource *certprotos.Resource) certprotos.Effect {
-	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
-		return certprotos.Effect_UNKNOWN
+func getTenantPolicyNetworkResources(ctx context.Context, policy *certprotos.Policy) ([]*certprotos.PolicyResource, error) {
+	var networkResources []*certprotos.PolicyResource
+	for _, policyResource := range policy.Resources.Resources {
+		if t := policyResource.GetTenant(); t != nil {
+			var networks []string
+			for _, i := range t.Tenants {
+				tenant, err := tenants.GetTenant(ctx, i)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "%v", err)
+				}
+				networks = append(networks, tenant.Networks...)
+			}
+
+			resource := &certprotos.PolicyResource{
+				Effect:   policyResource.Effect,
+				Action:   policyResource.Action,
+				Resource: &certprotos.PolicyResource_Network{Network: &certprotos.NetworkResource{Networks: networks}},
+			}
+			networkResources = append(networkResources, resource)
+		}
 	}
-	reqPath := reqResource.Resource
-	policyPath := policyResource.Resource
-	if ok, _ := doublestar.Match(policyPath, reqPath); ok {
-		return policyResource.Effect
+	return networkResources, nil
+}
+
+func getRequestResourceType(reqResource *certprotos.RequestResource) constants.ResourceType {
+	reqResourceType := constants.Path
+	switch reqResource.ResourceId.(type) {
+	case *certprotos.RequestResource_NetworkId:
+		reqResourceType = constants.NetworkID
+	case *certprotos.RequestResource_TenantId:
+		reqResourceType = constants.TenantID
 	}
-	return certprotos.Effect_UNKNOWN
+	return reqResourceType
+}
+
+func getPolicyResourceType(policyResource *certprotos.PolicyResource) constants.ResourceType {
+	policyResourceType := constants.Path
+	switch policyResource.Resource.(type) {
+	case *certprotos.PolicyResource_Network:
+		policyResourceType = constants.NetworkID
+	case *certprotos.PolicyResource_Tenant:
+		policyResourceType = constants.TenantID
+	}
+	return policyResourceType
+}
+
+// checkResourceType checks if the policy type applies to the requested type
+func checkResourceType(reqResource *certprotos.RequestResource, policyResource *certprotos.PolicyResource) bool {
+	reqResourceType := getRequestResourceType(reqResource)
+	policyResourceType := getPolicyResourceType(policyResource)
+	return reqResource.GetResourceId() != nil && policyResourceType != constants.Path && reqResourceType != policyResourceType
```

Comment:
My understanding was that given a policy where the user has read access to everything and denied write access to a specific network, if the user were to make a request with a random network `test`, s/he could still read `test` based on the first resource of the policy, i.e., the requested resource type here would be `network_id_type` but because there is no policy resource matching for this type and this specific resource id, we would have to match based on the first part of the policy. So the resource type matching for a given policyResource and requestedResource would only matter when the resource is of type network or tenant, thus the `!= constants.Path`
```
{
	Effect:   certprotos.Effect_ALLOW,
	Action:   certprotos.Action_READ,
	Resource: &certprotos.PolicyResource_Path{Path: &certprotos.PathResource{Path: "**"}},
},
{
	Effect:   certprotos.Effect_DENY,
	Action:   certprotos.Action_WRITE,
	Resource: &certprotos.PolicyResource_Network{Network: &certprotos.NetworkResource{Networks: []string{WriteTestNetworkId}}},
},
```